rust/main/agents/relayer/src/server/message_retry.rs

```diff
@@ -1,32 +1,95 @@
 use crate::settings::matching_list::MatchingList;
+
 use axum::{extract::State, routing, Json, Router};
+
 use derive_new::new;
-use tokio::sync::broadcast::Sender;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{broadcast::Sender, mpsc};
 
 const MESSAGE_RETRY_API_BASE: &str = "/message_retry";
 
-#[derive(new, Clone)]
+#[derive(new)]
 pub struct MessageRetryApi {
-    tx: Sender<MatchingList>,
+    retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryApiState {
+    pub retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryRequest {
+    pub uuid: String,
+    pub pattern: MatchingList,
+    pub transmitter: mpsc::Sender<MessageRetryResponse>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct MessageRetryResponse {
+    /// ID of the retry request
+    pub uuid: String,
+    /// how many pending operations were evaluated
+    pub evaluated: usize,
+    /// how many of the pending operations matched the retry request pattern
+    pub matched: u64,
 }
 
 async fn retry_message(
-    State(tx): State<Sender<MatchingList>>,
+    State(state): State<MessageRetryApiState>,
     Json(retry_req_payload): Json<MatchingList>,
-) -> String {
-    match tx.send(retry_req_payload) {
-        Ok(_) => "Moved message(s) to the front of the queue".to_string(),
-        // Technically it's bad practice to print the error message to the user, but
-        // this endpoint is for debugging purposes only.
-        Err(err) => format!("Failed to send retry request to the queue: {}", err),
+) -> Result<Json<MessageRetryResponse>, String> {
+    let uuid = uuid::Uuid::new_v4();
+    let uuid_string = uuid.to_string();
+
+    tracing::debug!(uuid = uuid_string, "Sending message retry request");
+
+    // This channel is only created to service this single
+    // retry request so we're expecting a single response
```

Comment:
as we expect 1 response from each transmitter end, if there are N transmitter ends (i.e. N submission tasks), wouldn't we expect N responses sent via the channel? So this would cause all sends other than the first one to block?

Reply:
my bad, I had thought that this channel was used to notify recipients of the retry request for some reason (so single producer, multiple consumer), and suggested to jeff to set this to 1: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/5043/files/a32c8dd0507db672fb1165ffcd5127a56f1bf17a#r1899741152

agree that with this being the channel used by submitters to send messages to the server it should have the length of the configured `relayChains`

Reply:
sounds good, I updated it to create a channel same size as `originChains.len()`